### ISSUE TITLE
Unknown Type Warning for Variable in The Conditional Assignment

### ISSUE THREAD

**Issue by syncoding** [microsoft/pyright#9556]
## Environment data

-   Pylance version: v2024.12.1
-   OS and version: Windows 10.0.19045
-   Python version (& distribution if applicable, e.g. Anaconda): 3.12.7 - Conda

## Code Snippet

```python
    @staticmethod
    def _alter_new(cmp: Company, data: dict[str, str | Real | None]) -> Company:
        cmp.fundamentals.current.quarter = (
            val if isinstance(val := data.get(SCR_STRUCT.data.quarter), str) else "Unknown"
        )
        cmp.fundamentals.current.balance_sheet.current_assets = (
            val if valid_number(val := data.get(SCR_STRUCT.data.current_assets)) else None  # valid_numer is a TypeGuard of Real
        )
        cmp.fundamentals.current.balance_sheet.financial_investments = (
            val if valid_number(val := data.get(SCR_STRUCT.data.financial_investments)) else None
        )
        .
        .
        .
        (more than100 attributes of the class are assigned with the same way using 'val' placeholder for type-check)
```

## Repro Steps

 

1. Behaviour starts after more than 100 attribute assignments are done with one under the other in the same method.

## Expected behavior

Both `val` variables in the individual assignment statements must be OK

## Actual behavior

Type of "val" is unknown Pylance(reportUnknownVariableType) warning is given for the `val` variables before the walrus operators.

## Logs

```
XXX
```

**Comment by rchiodo** [comment]
My guess is you're hitting some complexity limit but @erictraut will know more. Transferring to Pyright.

Can you include a full sample that reproduces the problem?

**Comment by erictraut** [comment]
This is unlikely to be a bug in pyright, but there isn't sufficient information here to say for sure.

If you can provide a minimal, self-contained sample that reproduces the problem, I'm happy to help you address the issue.

Closing the issue for now.

**Comment by syncoding** [comment]
When I change the variable name from `val` to anything else (say x), Pylance doesn't complain with this assignment statement, by the way. I mean it just complains with the variable name `val` after it's used a lot of time.

**Comment by syncoding** [comment]
And below trivial example gives the same warning for the varible ´val´ just for the one before the walrus operator. 

```python
   from random import choice

x = choice(["x", 1])
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
val if isinstance(val := x, str) else None
```

**Comment by erictraut** [comment]
You've created a block of code with high cyclomatic complexity. The code flow graph has more than 80 independent conditionals. While the Python interpret will pick only one path for each at runtime, a static analyzer needs to analyze all possible paths through the code.

Pyright has a number of internal limits to prevent poor performance and hangs, and you're hitting one of those with this code. If pyright didn't include such limits, this code would require 2^80 passes to compute all possible paths. This would never complete.

I presume that you have "strict" type checking mode enabled because the `reportUnkonwnVariableType` check isn't enabled by default in "standard" type checking mode.

If you want your code to be statically analyzable, you'll need to simplify it. Perhaps break it into subroutines or use a `match` statement or an `if`/`elif` ladder rather than independent `if`/`else` statements.

**Comment by syncoding** [comment]
Thank you for your kind explanation about the internals and pathways. (Type checking mode is set to `strict` by the way)
